rig/machine_control/machine_controller.py

```diff
@@ -0,0 +1,1193 @@
+        self._send_scp(x, y, p, SCPCommands.write, address, length_bytes,
+                       int(data_type), data, expected_args=0)
+
+    @ContextMixin.use_contextual_arguments
+    def read(self, address, length_bytes, x=Required, y=Required, p=0):
+        """Read a bytestring from an address in memory.
+
+        Parameters
+        ----------
+        address : int
+            The address at which to start reading the data.
+        length_bytes : int
+            The number of bytes to read from memory.
+
+        Returns
+        -------
+        :py:class:`bytes`
+            The data is read back from memory as a bytestring.
+        """
+        # Make calls to the lower level read method until we have read
+        # sufficient bytes.
+        data = b''
+        while len(data) < length_bytes:
+            # Determine the number of bytes to read
+            reads = min(self.scp_data_length, length_bytes - len(data))
+
+            # Determine the data type to use
+            dtype = address_length_dtype[(address % 4, reads % 4)]
+
+            # Perform the read and increment the address
+            data += self._read(x, y, p, address, reads, dtype)
+            address += reads
+
+        return data
+
+    def _read(self, x, y, p, address, length_bytes, data_type=DataType.byte):
+        """Read an SCP command's worth of data from an address in memory.
+
+        It is better to use :py:func:`~.read` which wraps this method and
+        allows reading bytestrings of arbitrary length.
+
+        Parameters
+        ----------
+        address : int
+            The address at which to start reading the data.
+        length_bytes : int
+            The number of bytes to read from memory, must be <=
+            :py:attr:`.scp_data_length`
+        data_type : DataType
+            The size of the data to write into memory.
+
+        Returns
+        -------
+        :py:class:`bytes`
+            The data is read back from memory as a bytestring.
+        """
+        read_scp = self._send_scp(x, y, p, SCPCommands.read, address,
+                                  length_bytes, int(data_type),
+                                  expected_args=0)
+        return read_scp.data
+
+    @ContextMixin.use_contextual_arguments
+    def read_struct_field(self, struct_name, field_name,
+                          x=Required, y=Required, p=0):
+        """Read the value out of a struct maintained by SARK.
+
+        Parameters
+        ----------
+        struct_name : string
+            Name of the struct to read from, e.g., `"sv"`
+        field_name : string
+            Name of the field to read, e.g., `"eth_addr"`
+
+        .. note::
+            The value returned is unpacked given the struct specification.
+
+        .. warning::
+            This feature is only available if this machine controller was used
+            to boot a board OR an appropriate struct definition has been
+            provided.  To do this use, e.g::
```

Comment:
Do you think it reasonable that perhaps the MachineController should accept a set of structs as an argument which defaults to simply loading the same struct file the boot command defaults to loading? 99.948% of people will never want any different and hopefully the number of people whose scripts will start with "boot the board" should gradually tend towards 0 over time. (/me is happy to add this quickly).